scripts/release.py

```diff
@@ -0,0 +1,276 @@
+        for name, email in unknowns:
+            commit = self.__modin_repo.get_commit(str(email2pr[email]))
+            found = set()
+            for pull in commit.get_pulls():
+                for review in pull.get_reviews():
+                    user = review.user
+                    if user.name == name and (not user.email or user.email == email):
+                        found.add(user.login)
+
+            if len(found) == 1:
+                self.__register(name, email, list(found)[0])
+                logins |= found
+            else:
+                new_unknowns.add((name, email))
+
+        return logins, new_unknowns
+
+    def __save(self):
+        self.__cache_file.write_text(json.dumps(self.__cache, indent=4, sort_keys=True))
+
+
+class GitWrapper:
+    def __init__(self):
+        self.repo = pygit2.Repository(Path(__file__).parent)
+
+    def is_on_master(self):
+        return self.repo.references["refs/heads/master"] == self.repo.head
+
+    def get_previous_release(self, rel_type):
+        tags = [
+            (entry, version.parse(entry.lstrip("refs/tags/")))
+            for entry in self.repo.references
+            if entry.startswith("refs/tags/")
+        ]
+        # filter away legacy versions (which aren't following the proper naming schema)
+        tags = [(entry, ver) for entry, ver in tags if isinstance(ver, version.Version)]
+        if rel_type == "minor":
+            # leave only minor releases
+            tags = [(entry, ver) for entry, ver in tags if ver.micro == 0]
+        else:
+            assert rel_type == "patch"
+        prev_ref, prev_ver = max(tags, key=lambda pair: pair[1])
+        return prev_ref, self.repo.references[prev_ref].peel(), prev_ver
+
+    def get_commits_upto(self, stop_commit):
+        history = []
+        for obj in self.repo.walk(self.repo.head.target):
+            if obj.id == stop_commit.id:
+                break
+            history.append(obj)
+        else:
+            raise ValueError("Current HEAD is not derived from previous release")
+        return history
+
+    def ensure_title_link(self, obj: pygit2.Commit):
+        title = obj.message.splitlines()[0]
+        if not re.match(r".*\(#(\d+)\)$", title):
+            title += f" ({obj.short_id})"
+        return title
+
+
+def make_notes(args):
+    wrapper = GitWrapper()
+    release_type = "minor" if wrapper.is_on_master() else "patch"
+    sys.stderr.write(f"Detected release type: {release_type}\n")
+
+    prev_ref, prev_commit, prev_ver = wrapper.get_previous_release(release_type)
+    sys.stderr.write(f"Previous {release_type} release: {prev_ref}\n")
+
+    next_major, next_minor, next_patch = prev_ver.release
+    if release_type == "minor":
+        next_minor += 1
+    elif release_type == "patch":
+        next_patch += 1
+    else:
+        raise ValueError(f"Unexpected release type: {release_type}")
+    next_ver = version.Version(f"{next_major}.{next_minor}.{next_patch}")
+
```

Comment:
For now we only do `minor` and `patch`. When we would start doing major, I would assume this would still happen very rarely, and by the time we'd be ready for 1.0 we'll fix the script to support that 🙃